waspc/src/Wasp/Generator/Monad.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.Generator.Monad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    catchGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the error that halted generation. Every error is fatal.
+-- There may optionally be additional errors or non-fatal warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError = GenericGeneratorError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+
+data GeneratorWarning = GenericGeneratorWarning String
+
+instance Show GeneratorWarning where
+  show (GenericGeneratorWarning e) = e
+
+-- The outer Either layer represents the error that halted generation. Every error is fatal.
+-- In either case, we return a list of non-fatal warnings and the list of all errors encountered.
+-- However, the implication is any error will cause generation to fail,
+-- and the top-level Either error is also included in the list of all errors encountered.
+runGenerator :: Generator a -> (Either GeneratorError a, ([GeneratorWarning], [GeneratorError]))
```

Comment:
What about
```hs
runGenerator :: Generator a -> ([GeneratorWarning], Either [GeneratorError] a)
```
or if we wanted to be more fancy:
```hs
runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
```

This feels to me like it matches closer what we want to convey: Once generator runs, we want to get all the warnings (0 or more), and we want to get either generated file drafts, if there were no errors, or get all the errors, if there were errors, and in that case we don't need file drafts since we won't be writing anything (what is the purpose of returning file drafts next to errors -> not much, since we don't know what happens if we write them).

Reply:
Good idea, I think `runGenerator :: Generator a -> ([GeneratorWarning], Either [GeneratorError] a)` captures it well to me. Will update now, thanks! 👍🏻

Reply:
Great! What about using `NonEmpty`? That way we avoid the weird state of failing with no errors (empty list).

Reply:
Haven't used it before, but from the name I'm guess non-empty List? :D That is probably a better way to convey that requirement. Let me give it a try!

Reply:
I was able to update it to use `NonEmpty`, but I still had to rely on the fact that if `result` is a `Left GeneratorError`, it means _at least that error_ should also be in the errors of state:

```haskell
runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
runGenerator generator =
  let (result, state') = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
   in (warnings state', left (\_ -> fromList $ errors state') result)
  where
    state = GeneratorState {warnings = [], errors = []}
```

That invariant should be upheld with `logAndThrowGeneratorError`, however. Is this the sort of change you had in mind? https://github.com/wasp-lang/wasp/pull/433/commits/daabe8f94fd5ede15bd5c3cc0383c05df532d718

Reply:
I see what you mean, that is an invariant indeed, and as you said, it should be enforced by the internal workings of the Generator Monad.

However I wanted to achieve another thing by changing this type -> the idea was that user of `runGenerator` can be sure that `runGenerator` won't return an empty list of errors when it fails -> it will be either a list with at least one error, or file drafts. This way it doesn't have to check the weird case when there are no errors.

Reply:
Yep, I agree completely on the benefits from the caller side. It was an internal implied invariant before (not great), and now the type system ensures it from the caller perspective (must better)!